sky/templates/kubernetes-ray.yml.j2

```diff
@@ -616,7 +651,49 @@ available_node_types:
               {{k8s_resource_key}}: {{accelerator_count}}
               {% endif %}
             {% endif %}
-            
+      
+      {% if high_availability %}
+      pvc_spec:
+        apiVersion: v1
+        kind: PersistentVolumeClaim
+        metadata:
+          name: {{cluster_name_on_cloud}}-{{k8s_high_availability_deployment_volume_mount_name}}
+          namespace: {{k8s_namespace}}
+        spec:
+          accessModes:
+            - ReadWriteOnce # Our controller pod is singleton
+          resources:
+            requests:
+              storage: {{disk_size}}
+
+      deployment_spec:
+        apiVersion: apps/v1
+        kind: Deployment
+        metadata:
+          name: {{cluster_name_on_cloud}}-deployment
+          namespace: {{k8s_namespace}}
+        spec:
+          replicas: 1
+          selector:
+            matchLabels:
+              app: {{cluster_name_on_cloud}}
+          template:
+            # The only difference between the pod spec and this section is the initContainers
+            metadata:
+              # should be replaced by pod metadata
+            spec:
+              # To prevent the home dir provided by the docker image from being overriden by pvc mounting,
+              # we use initContainers to copy it first to /mnt/home, which will later be mounted to home dir.
+              initContainers:
+                - name: init-copy-home
+                  image: {{image_id}}
+                  command: ["/bin/sh", "-c"]
+                  args: ["echo 'Copying home directory to /mnt/home'; rsync -a {{k8s_high_availability_deployment_volume_mount_path}}/ /mnt/home && echo 'Copy completed.'; ls -la /mnt/home"]
+                  volumeMounts:
+                    - name: {{k8s_high_availability_deployment_volume_mount_name}}
+                      mountPath: /mnt/home
```

Comment:
What if the Pod get recreated? Looks like this will override the persistent home path with the new home path of the image when the init-container get executed again.

Reply:
Exactly. While rsync might update files that came from the image if they differ, the key benefit here is that it doesn't delete files created solely within the PVC, preserving user data across restarts. But you're correct, running it every time does add some overhead. Adding a check for an initialization marker is a great optimization to avoid this on subsequent Pod starts. Let's implement that check!

Reply:
That's a fair point. You're right, we shouldn't implicitly rely on the controller avoiding modifications to files that also exist in the image. That assumption isn't inherently safe or robust.